nemo/collections/nlp/models/language_modeling/megatron/falcon/falcon_decoder_layer.py

```diff
@@ -0,0 +1,284 @@
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask,
+        context=None,
+        context_mask=None,
+        rotary_pos_emb=None,
+        inference_params=None,
+    ):
+        # hidden_states: [s, b, h]
+
+        # Residual connection.
+        residual = hidden_states
+
+        if self.new_decoder_architecture:
+            mlp_ln_output = self.pre_mlp_layernorm(hidden_states)
+
+        # Optional Input Layer norm
+        input_layernorm_output = self.input_layernorm(hidden_states)
+
+        input_mlp_ln = input_layernorm_output
+
+        # Self attention.
+        attention_output_with_bias = self.self_attention(
+            input_layernorm_output,
+            attention_mask=attention_mask,
+            inference_params=inference_params,
+            rotary_pos_emb=rotary_pos_emb,
+        )
+
+        # TODO: could we move `bias_dropout_add_exec_handler` itself
+        # inside the module provided in the `bias_dropout_add_spec` module?
+        with self.bias_dropout_add_exec_handler():
+            hidden_states = self.self_attn_bda(self.training, self.config.bias_dropout_fusion)(
+                attention_output_with_bias, residual, self.config.hidden_dropout
+            )
+
+        if not self.new_decoder_architecture:
+            if self.parallel_attention:
+                layernorm_output = input_mlp_ln
+            else:
+                residual = hidden_states
+                layernorm_output = self.post_self_attn_layernorm(hidden_states)
+
+        else:
+            layernorm_output = mlp_ln_output
+
+        mlp_output_with_bias = self.mlp(layernorm_output)
+
+        # falcon specific:
+        if self.new_decoder_architecture or self.parallel_attention:
+            mlp_output = mlp_output_with_bias[0]
+            attn_output = attention_output_with_bias[0]
+            mlp_output_without_bias = mlp_output + attn_output
+            mlp_output_with_bias = (mlp_output_without_bias, None)
+
+        # TODO: could we move `bias_dropout_add_exec_handler` itself
+        # inside the module provided in the `bias_dropout_add_spec` module?
+        with self.bias_dropout_add_exec_handler():
+            hidden_states = self.mlp_bda(self.training, self.config.bias_dropout_fusion)(
+                mlp_output_with_bias, residual, self.config.hidden_dropout
+            )
+
+        # Jit compiled function creates 'view' tensor. This tensor
+        # potentially gets saved in the MPU checkpoint function context,
+        # which rejects view tensors. While making a viewless tensor here
+        # won't result in memory savings (like the data loader, or
+        # p2p_communication), it serves to document the origin of this
+        # 'view' tensor.
+        output = make_viewless_tensor(inp=hidden_states, requires_grad=hidden_states.requires_grad, keep_graph=True)
+
+        return output, context
+
+    def sharded_state_dict(self, prefix=''):
```

Comment:
This doesn't have anything falcon specific? Can we reuse the original code somehow?

Reply:
per previous conversation, will make a optimized version for sharded_state_dict in future PR.